File: apps/web/src/lib/bootstrapStreamModels.ts
import type { EntryListItem } from "@/lib/atprotoClient";
import type { PublicationSidebarProjection } from "@/lib/publicationProjectionClient";

export type BootstrapStreamEventKind =
  | "sidebarPriority"
  | "unreadCounts"
  | "selectedPublication"
  | "entriesPage"
  | "sidebarFolders"
  | "warning"
  | "error"
  | "done";

export type BootstrapStreamEvent = {
  kind: BootstrapStreamEventKind;
  sidebarPriority?: PublicationSidebarProjection;
  unreadCounts?: { counts: Record<string, number> };
  selectedPublication?: { publicationId: string };
  entriesPage?: {
    publicationId: string;
    entries: EntryListItem[];
    cursor?: string;
  };
  sidebarFolders?: {
    folderSections: NonNullable<PublicationSidebarProjection["folderSections"]>;
    allPublicationRows: PublicationSidebarProjection["allPublicationRows"];
  };
  warning?: { message: string };
  error?: { message: string };
  done?: { refreshedAt: string };
};

export type ParsedBootstrapStreamEvent =
  | { kind: "sidebarPriority"; payload: PublicationSidebarProjection }
  | { kind: "unreadCounts"; payload: { counts: Record<string, number> } }
  | { kind: "selectedPublication"; payload: { publicationId: string } }
  | {
      kind: "entriesPage";
      payload: {
        publicationId: string;
        entries: EntryListItem[];
        cursor?: string;
      };
    }
  | {
      kind: "sidebarFolders";
      payload: {
        folderSections: NonNullable<PublicationSidebarProjection["folderSections"]>;
        allPublicationRows: PublicationSidebarProjection["allPublicationRows"];
      };
    }
  | { kind: "warning"; payload: { message: string } }
  | { kind: "error"; payload: { message: string } }
  | { kind: "done"; payload: { refreshedAt: string } };

export function parseBootstrapStreamEvent(
  raw: BootstrapStreamEvent
): ParsedBootstrapStreamEvent | null {
  switch (raw.kind) {
    case "sidebarPriority":
      if (!raw.sidebarPriority) return null;
      return { kind: "sidebarPriority", payload: raw.sidebarPriority };
    case "unreadCounts":
      if (!raw.unreadCounts) return null;
      return { kind: "unreadCounts", payload: raw.unreadCounts };
    case "selectedPublication":
      if (!raw.selectedPublication) return null;
      return { kind: "selectedPublication", payload: raw.selectedPublication };
    case "entriesPage":
      if (!raw.entriesPage) return null;
      return { kind: "entriesPage", payload: raw.entriesPage };
    case "sidebarFolders":
      if (!raw.sidebarFolders) return null;
      return { kind: "sidebarFolders", payload: raw.sidebarFolders };
    case "warning":
      if (!raw.warning) return null;
      return { kind: "warning", payload: raw.warning };
    case "error":
      if (!raw.error) return null;
      return { kind: "error", payload: raw.error };
    case "done":
      if (!raw.done) return null;
      return { kind: "done", payload: raw.done };
    default:
      return null;
  }
}

export function firstUnreadPriorityPublicationId(args: {
  myPublications: PublicationSidebarProjection["myPublications"];
  subscribedUnfoldered: PublicationSidebarProjection["subscribedUnfoldered"];
  followingTabPublications: PublicationSidebarProjection["followingTabPublications"];
  unreadCounts: Record<string, number>;
}): string | null {
  for (const row of [
    ...args.myPublications,
    ...args.subscribedUnfoldered,
    ...args.followingTabPublications,
  ]) {
    const count = args.unreadCounts[row.publicationId] ?? row.unreadCount ?? 0;
    if (count > 0) return row.publicationId;
  }
  return null;
}

File: packages/swift/GatewayCore/Sources/GatewayCore/Models/BootstrapStreamModels.swift
import Foundation

public enum AppViewBootstrapStreamEventKind: String, Codable, Sendable {
  case sidebarPriority
  case unreadCounts
  case selectedPublication
  case entriesPage
  case sidebarFolders
  case warning
  case error
  case done
}

public struct AppViewBootstrapUnreadCountsPayload: Codable, Sendable, Equatable {
  public let counts: [String: Int]

  public init(counts: [String: Int]) {
    self.counts = counts
  }
}

public struct AppViewBootstrapSelectedPublicationPayload: Codable, Sendable, Equatable {
  public let publicationId: String

  public init(publicationId: String) {
    self.publicationId = publicationId
  }
}

public struct AppViewBootstrapEntriesPagePayload: Codable, Sendable, Equatable {
  public let publicationId: String
  public let entries: [AppViewBootstrapEntryListItem]
  public let cursor: String?

  public init(publicationId: String, entries: [AppViewBootstrapEntryListItem], cursor: String?) {
    self.publicationId = publicationId
    self.entries = entries
    self.cursor = cursor
  }
}

public struct AppViewBootstrapEntryListItem: Codable, Sendable, Equatable {
  public let entryId: String
  public let title: String
  public let summary: String?
  public let publishedAt: Date
  public let thumbnailUrl: String?
  public let thumbnailFallbackUrl: String?

  public init(
    entryId: String,
    title: String,
    summary: String? = nil,
    publishedAt: Date,
    thumbnailUrl: String? = nil,
    thumbnailFallbackUrl: String? = nil
  ) {
    self.entryId = entryId
    self.title = title
    self.summary = summary
    self.publishedAt = publishedAt
    self.thumbnailUrl = thumbnailUrl
    self.thumbnailFallbackUrl = thumbnailFallbackUrl
  }
}

public struct AppViewBootstrapSidebarFoldersPayload: Codable, Sendable, Equatable {
  public let folderSections: [PublicationFolderSection]
  public let allPublicationRows: [SidebarPublicationRow]

  public init(folderSections: [PublicationFolderSection], allPublicationRows: [SidebarPublicationRow]) {
    self.folderSections = folderSections
    self.allPublicationRows = allPublicationRows
  }
}

public struct AppViewBootstrapMessagePayload: Codable, Sendable, Equatable {
  public let message: String

  public init(message: String) {
    self.message = message
  }
}

public struct AppViewBootstrapDonePayload: Codable, Sendable, Equatable {
  public let refreshedAt: Date

  public init(refreshedAt: Date) {
    self.refreshedAt = refreshedAt
  }
}

public struct AppViewBootstrapStreamEvent: Codable, Sendable {
  public let kind: AppViewBootstrapStreamEventKind
  public let sidebarPriority: PublicationSidebarResponse?
  public let unreadCounts: AppViewBootstrapUnreadCountsPayload?
  public let selectedPublication: AppViewBootstrapSelectedPublicationPayload?
  public let entriesPage: AppViewBootstrapEntriesPagePayload?
  public let sidebarFolders: AppViewBootstrapSidebarFoldersPayload?
  public let warning: AppViewBootstrapMessagePayload?
  public let error: AppViewBootstrapMessagePayload?
  public let done: AppViewBootstrapDonePayload?

  public init(
    kind: AppViewBootstrapStreamEventKind,
    sidebarPriority: PublicationSidebarResponse? = nil,
    unreadCounts: AppViewBootstrapUnreadCountsPayload? = nil,
    selectedPublication: AppViewBootstrapSelectedPublicationPayload? = nil,
    entriesPage: AppViewBootstrapEntriesPagePayload? = nil,
    sidebarFolders: AppViewBootstrapSidebarFoldersPayload? = nil,
    warning: AppViewBootstrapMessagePayload? = nil,
    error: AppViewBootstrapMessagePayload? = nil,
    done: AppViewBootstrapDonePayload? = nil
  ) {
    self.kind = kind
    self.sidebarPriority = sidebarPriority
    self.unreadCounts = unreadCounts
    self.selectedPublication = selectedPublication
    self.entriesPage = entriesPage
    self.sidebarFolders = sidebarFolders
    self.warning = warning
    self.error = error
    self.done = done
  }

  public static func sidebarPriority(_ response: PublicationSidebarResponse) -> Self {
    Self(kind: .sidebarPriority, sidebarPriority: response)
  }

  public static func unreadCounts(_ counts: [String: Int]) -> Self {
    Self(kind: .unreadCounts, unreadCounts: .init(counts: counts))
  }

  public static func selectedPublication(publicationId: String) -> Self {
    Self(kind: .selectedPublication, selectedPublication: .init(publicationId: publicationId))
  }

  public static func entriesPage(_ payload: AppViewBootstrapEntriesPagePayload) -> Self {
    Self(kind: .entriesPage, entriesPage: payload)
  }

  public static func sidebarFolders(_ payload: AppViewBootstrapSidebarFoldersPayload) -> Self {
    Self(kind: .sidebarFolders, sidebarFolders: payload)
  }

  public static func warning(_ message: String) -> Self {
    Self(kind: .warning, warning: .init(message: message))
  }

  public static func error(_ message: String) -> Self {
    Self(kind: .error, error: .init(message: message))
  }

  public static func done(refreshedAt: Date) -> Self {
    Self(kind: .done, done: .init(refreshedAt: refreshedAt))
  }
}

public enum AppViewBootstrapStreamNDJSON {
  public static func encodeLine(_ event: AppViewBootstrapStreamEvent) throws -> Data {
    let encoder = JSONEncoder()
    encoder.dateEncodingStrategy = .iso8601
    var data = try encoder.encode(event)
    data.append(0x0A)
    return data
  }
}
